metricbeat/beater/reload.go

```diff
@@ -0,0 +1,221 @@
+	gcd := NewGlobalChangeDetector(r.config.Path)
+
+	for {
+		select {
+		case <-r.done:
+			logp.Info("Dynamic config reloader stopped")
+			return
+		case <-time.After(r.config.Period):
+
+			debugr("Scan for new config files")
+			configReloads.Add(1)
+
+			files, updated, err := gcd.Scan()
+			if err != nil {
+				logp.Err("Error fetching new config files: %v", err)
+				// Continue on error or no new files
+				continue
+			}
+
+			// no file changes
+			if !updated {
+				continue
+			}
+
+			// Load all config objects
+			configs := []*common.Config{}
+			for _, file := range files {
+				c, err := LoadConfigs(file)
+				if err != nil {
+					logp.Err("Error loading config: %s", err)
+					continue
+				}
+
+				configs = append(configs, c...)
+			}
+
+			// Check which configs do not exist anymore
+			s, err := NewModuleWrappers(configs, mb.Registry)
+			if err != nil {
+				if err != mb.ErrAllModulesDisabled && err != mb.ErrEmptyConfig {
+					// Continuing as only some modules could have an error
+					logp.Err("Error creating modules: %s", err)
+				}
+			}
+
+			debugr("Number of module wrappers created: %v", len(s))
+
+			startList := []*ModuleWrapper{}
+			stopList := r.registry.CopyList()
+
+			for _, w := range s {
+
+				// Only add modules to startlist which are enabled
+				if !w.Config().Enabled {
+					continue
+				}
+
+				debugr("Remove from stoplist: %v", w.Hash())
+				delete(stopList, w.Hash())
+
+				// As module already exist, it must be removed from the stop list and not started
+				if !r.registry.Has(w.Hash()) {
+					debugr("Add to startlist: %v", w.Hash())
+					startList = append(startList, w)
+					continue
+				}
+			}
+
+			r.StopModules(stopList)
+			r.StartModules(startList)
+		}
+	}
+}
+
+func (r *ConfigReloader) Stop() {
+	close(r.done)
+	r.wg.Wait()
+}
+
+func (r *ConfigReloader) StartModules(list []*ModuleWrapper) {
+
+	logp.Info("Starting %v modules ...", len(list))
+	for _, w := range list {
+		go func(mw *ModuleWrapper) {
+			mr := NewModuleRunner(r.client, mw)
+			mr.Start()
+			r.registry.Add(mw.Hash(), mr)
+			moduleStarts.Add(1)
+			moduleRunning.Add(1)
+			debugr("New Module Started: %v", mw.Hash())
+		}(w)
+	}
+}
+
+func (r *ConfigReloader) StopModules(list map[uint64]ModuleRunner) {
+	logp.Info("Stopping %v modules ...", len(list))
+
+	for hash, w := range list {
+		w.Stop()
+		r.registry.Remove(hash)
+		moduleStops.Add(1)
+		moduleRunning.Add(-1)
+		debugr("Module stopped: %v", hash)
+	}
+}
+
+// LoadConfigs loads the configs data from the given file
+func LoadConfigs(file string) ([]*common.Config, error) {
+	debugr("Load config from file: %s", file)
+	rawConfig, err := common.LoadFile(file)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid config: %s", err)
+	}
+
+	var c []*common.Config
+	err = rawConfig.Unpack(&c)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading configuration from file %s: %s", file, err)
+	}
+
+	return c, nil
+}
+
+func NewRunningRegistry() RunningRegistry {
```

Comment:
Given that this struct contains a `sync.Mutex`, returning a `*RunningRegistry` would better indicate to the caller that this object needs to be passed by reference than by value (as you should not copy a `sync.Mutex`).